metadata/metadata.go

```diff
@@ -0,0 +1,159 @@
+// and encoding format -- it's a protocol -- or maybe we're ultimately just
+// describing a metadata encoding format?
+const DataTransferProtocolID = 0x300001
+
+// ExchangeFormat identifies a protocol for exchanging vouchers over data transfer
+type ExchangeFormat string
+
+// TransportProtocol indicates the underlying transport data transfer will use
+// Note: while currently go-data-transfer supports only graphsync, this is part
+// of the metadata to future proof for multiple transport protocols
+type TransportProtocol string
+
+// DataTransferMetadata is the CBOR serialized data encoded in the
+// metadata bytes when the multicodec is DataTransferProtocolID
+type DataTransferMetadata struct {
+	// ExchangeFormat identifies the type of vouchers taht will be exchanged
+	ExchangeFormat ExchangeFormat
+	// Transport indicates the underlying protocol (for now, always graphsync)
+	TransportProtocol TransportProtocol
+	// FormatData is opaque bytes that can be decoded once you know the format type
+	FormatData []byte
+}
+
+const (
+	// FilecoinV1 is the current filecoin retrieval protocol
+	FilecoinV1 ExchangeFormat = "/filecoin-exchange/1.0.0"
+
+	// GraphSync is the name of actual graphsync libp2p protocol
+	GraphSync TransportProtocol = "/ipfs/graphsync/1.0.0"
+)
+
+// FilecoinV1Data is the information encoded in FormatDat for FilecoinV1
+type FilecoinV1Data struct {
+	// PieceCID identifies the piece this data can be found in
+	PieceCID datamodel.Link
+	// Free indicates if the retrieval is free
+	IsFree bool
+	// FastRetrieval indicates whether the provider claims there is an unsealed copy
+	FastRetrieval bool
+}
+
+var dataTransferSchemaType schema.Type
+var filecoinvV1SchemaType schema.Type
+
+func init() {
+	ts := schema.TypeSystem{}
+	ts.Init()
+	ts.Accumulate(schema.SpawnString("String"))
+	ts.Accumulate(schema.SpawnBytes("Bytes"))
+	ts.Accumulate(schema.SpawnLink("Link"))
+	ts.Accumulate(schema.SpawnBool("Bool"))
+	ts.Accumulate(schema.SpawnStruct("DataTransferMetadata",
+		[]schema.StructField{
+			schema.SpawnStructField("ExchangeFormat", "String", false, false),
+			schema.SpawnStructField("TransportProtocol", "String", false, false),
+			schema.SpawnStructField("FormatData", "Bytes", false, false),
+		},
+		schema.SpawnStructRepresentationMap(nil),
+	))
+	ts.Accumulate(schema.SpawnStruct("FilecoinV1Data",
+		[]schema.StructField{
+			schema.SpawnStructField("PieceCID", "Link", false, false),
+			schema.SpawnStructField("IsFree", "Bool", false, false),
+			schema.SpawnStructField("FastRetrieval", "Bool", false, false),
+		},
+		schema.SpawnStructRepresentationMap(nil),
+	))
+	dataTransferSchemaType = ts.TypeByName("DataTransferMetadata")
+	filecoinvV1SchemaType = ts.TypeByName("FilecoinV1Data")
+}
+
+// EncodeDataTransferMetadata writes DataTransferMetadata to a byte stream
+func EncodeDataTransferMetadata(dtm *DataTransferMetadata, w io.Writer) error {
+	nd := bindnode.Wrap(dtm, dataTransferSchemaType)
```

Comment:
Assuming `DataTransferMetadata` represents a full `Metadata` type storetheindex expects, does this encoding write `varint` Protocol ID at offset `0` of the resulting metadata bytes similar to what [sotoretheindex](https://github.com/filecoin-project/storetheindex/blob/main/api/v0/metadata.go#L46) expects? 

In a related subject I wonder if it would be possible to reuse some of the primitives provided by storetheindex to do metadata encode/decode?

Reply:
Ah I see; so the bytes produced by the encode function above is what we would set as [`Metadata.Data`](https://github.com/filecoin-project/storetheindex/blob/main/api/v0/metadata.go#L27). In that case please ignore me.

Reply:
Store the index expects a [`v0.Metadata`](https://github.com/filecoin-project/storetheindex/blob/main/api/v0/metadata.go#L22-L28) type, not an encoded block of bytes.  So it seems like this should be wrapped in a `v0.Metadata` type.